presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java

```diff
@@ -0,0 +1,620 @@
+    {
+        return typeCoercer.coerceTypeBase(sourceType, resultTypeBase);
+    }
+
+    public ScalarFunctionImplementation getScalarFunctionImplementation(FunctionHandle functionHandle)
+    {
+        if (functionHandle.getCatalogSchemaName().equals(SESSION_NAMESPACE)) {
+            return ((SessionFunctionHandle) functionHandle).getScalarFunctionImplementation();
+        }
+        Optional<FunctionNamespaceManager<?>> functionNamespaceManager = getServingFunctionNamespaceManager(functionHandle.getCatalogSchemaName());
+        checkArgument(functionNamespaceManager.isPresent(), "Cannot find function namespace for '%s'", functionHandle.getCatalogSchemaName());
+        return functionNamespaceManager.get().getScalarFunctionImplementation(functionHandle);
+    }
+
+    public CompletableFuture<SqlFunctionResult> executeFunction(String source, FunctionHandle functionHandle, Page inputPage, List<Integer> channels)
+    {
+        Optional<FunctionNamespaceManager<?>> functionNamespaceManager = getServingFunctionNamespaceManager(functionHandle.getCatalogSchemaName());
+        checkState(functionNamespaceManager.isPresent(), format("FunctionHandle %s should have a serving function namespace", functionHandle));
+        return functionNamespaceManager.get().executeFunction(source, functionHandle, inputPage, channels, this);
+    }
+
+    public WindowFunctionSupplier getWindowFunctionImplementation(FunctionHandle functionHandle)
+    {
+        return builtInTypeAndFunctionNamespaceManager.getWindowFunctionImplementation(functionHandle);
+    }
+
+    public InternalAggregationFunction getAggregateFunctionImplementation(FunctionHandle functionHandle)
+    {
+        return builtInTypeAndFunctionNamespaceManager.getAggregateFunctionImplementation(functionHandle);
+    }
+
+    public BuiltInScalarFunctionImplementation getBuiltInScalarFunctionImplementation(FunctionHandle functionHandle)
+    {
+        return (BuiltInScalarFunctionImplementation) builtInTypeAndFunctionNamespaceManager.getScalarFunctionImplementation(functionHandle);
+    }
+
+    @VisibleForTesting
+    public List<SqlFunction> listOperators()
+    {
+        Set<QualifiedObjectName> operatorNames = Arrays.asList(OperatorType.values()).stream()
+                .map(OperatorType::getFunctionName)
+                .collect(toImmutableSet());
+
+        return builtInTypeAndFunctionNamespaceManager.listFunctions(Optional.empty(), Optional.empty()).stream()
+                .filter(function -> operatorNames.contains(function.getSignature().getName()))
+                .collect(toImmutableList());
+    }
+
+    public FunctionHandle resolveOperator(OperatorType operatorType, List<TypeSignatureProvider> argumentTypes)
+    {
+        try {
+            return resolveFunction(Optional.empty(), Optional.empty(), operatorType.getFunctionName(), argumentTypes);
+        }
+        catch (PrestoException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)